cmd/kubeadm/app/cmd/alpha/kubeconfig.go

```diff
@@ -81,29 +77,23 @@ func newCmdUserKubeConfig(out io.Writer) *cobra.Command {
 			}
 
 			// This call returns the ready-to-use configuration based on the defaults populated by flags
-			internalcfg, err := configutil.DefaultedInitConfiguration(initCfg, clusterCfg)
+			internalCfg, err := configutil.LoadOrDefaultInitConfiguration(cfgPath, initCfg, clusterCfg)
 			if err != nil {
 				return err
 			}
 
 			// if the kubeconfig file for an additional user has to use a token, use it
 			if token != "" {
-				return kubeconfigphase.WriteKubeConfigWithToken(out, internalcfg, clientName, token)
+				return kubeconfigphase.WriteKubeConfigWithToken(out, internalCfg, clientName, token)
 			}
 
 			// Otherwise, write a kubeconfig file with a generate client cert
-			return kubeconfigphase.WriteKubeConfigWithClientCert(out, internalcfg, clientName, organizations)
+			return kubeconfigphase.WriteKubeConfigWithClientCert(out, internalCfg, clientName, organizations)
 		},
 		Args: cobra.NoArgs,
 	}
 
-	// Add ClusterConfiguration backed flags to the command
-	cmd.Flags().StringVar(&clusterCfg.CertificatesDir, options.CertificatesDir, clusterCfg.CertificatesDir, "The path where certificates are stored")
-	cmd.Flags().StringVar(&clusterCfg.ClusterName, "cluster-name", clusterCfg.ClusterName, "Cluster name to be used in kubeconfig")
-
-	// Add InitConfiguration backed flags to the command
-	cmd.Flags().StringVar(&initCfg.LocalAPIEndpoint.AdvertiseAddress, options.APIServerAdvertiseAddress, initCfg.LocalAPIEndpoint.AdvertiseAddress, "The IP address the API server is accessible on")
-	cmd.Flags().Int32Var(&initCfg.LocalAPIEndpoint.BindPort, options.APIServerBindPort, initCfg.LocalAPIEndpoint.BindPort, "The port the API server is accessible on")
+	options.AddConfigFlag(cmd.Flags(), &cfgPath)
```

Comment:
@neolit123 Currently this command only accept local kubeadm config, should we also consider the config saved in cluster?

Reply:
thanks for the PR, i can have a deeper look next week.

that is a good question. we had previous discussions whether certain kubeadm commands should use the kubeadm-config CM. my take is that if a command has potential to *not operate* on the running cluster it should not fetch the CM.

for example, if the command is currently being run on a Node part of a cluster and if a kubeconfig to access the cluster is in the user home dir, the user might still want to create a kubeadm for a completely different cluster, but the command will use the configuration from the cluster, which might not be desired.

i had the same argument for the command "kubeadm config images...". "kubeadm alpha certs renew" on the other hand has CM fetch, but the idea of the command is to renew certs for this Node / cluster.

so my preference here would be the following:
- the user should feed --config and the command should fail without it.
- if the user wants to feed the --config for the current cluster they can fetch the ClusterConfiguration from the CM.

Reply:
I got it, thanks for you detailed clarification!